source/grammar/index.rst

```diff
@@ -1,12 +1,132 @@
+repetition is denoted by curly brackets ``+`` (one or more) and ``*`` (zero or more). An ellipses
+``...`` is used to match any series of token.
 
-.. literalinclude:: qasm3.g4
-   :language: antlr
-   :linenos:
\ No newline at end of file
+.. productionlist::
+    program: header generic_statement*
+    header: version? include*
+    version: "OPENQASM" real ";"
+    include: "include" id ( ".inc" | ".qasm" ) ";"
+    generic_statement: global_statement
+        :| statement
+    global_statement: subroutine_declaration
+        :| kernel_declaration
+        :| gate_definition
+        :| calibration
+    statement: expression_statement
+        :| declaration_statement
+        :| selection_statement
+        :| iteration_statement
+        :| selection_directive_statement
+        :| quantum_statement
+        :| timing_box
+        :| pragma
+    subroutine_declaration: "def" id args_declaration return_signature? program_block
+    args_declaration : "(" type_and_id_list ")"
+    type_and_id_list: ( classical_type_and_id_list | quantum_type_and_id_list )*
+    classical_type_and_id_list: ( declare_type association "," )* declare_type association
+    declare_type: dependent_type_declaration 
+        :| independent_type_specifier
+    dependent_type_declaration: dependent_type_specifier designator
+    dependent_type_specifier: "int" | "uint" | "angle" | "fixed" | "stretch"
+    designator: "[" expression "]"
+    independent_type_specifier: "bit" | "creg" | "bool" 
+        :| timing_type
+    association: ":" id
+    return_signature: "->" classical_declaration
+    constant_declaration: "const" ( dependent_type_declaration | independent_type ) assignment 
+    classical_declaration: ( dependent_type_declaration | independent_type ) assignment?
+    program_block: "{" ( program_block | statement ) "}"
+    kernel_declaration: "kernel" id classical_type_and_id_list return_signature? ";"
+    expression_statement: expression ";"
+        :| "return" expression ";
+    expression: primary_expression 
+        :| expression binaryop expression
+        :| unary_operator expression 
+        :| membership_test
+        :| expression '[' expression ']'
+        :| call "(" expression_list? ")"
+        :| expression incrementor
+        :| quantum_measurement
+        :| expression_terminator
+    expression_terminator: real | nninteger | id
+        :| constant
+        :| time_terminator
+    constant: "pi" | "π" | "tau" | "𝜏" | "euler"
+    expression_list: ( expression "," )* expression 
+    unary_operator: "~" | "!"
+    call: expression
+        :| builtin_math
+        :| cast_operator
+    builtin_math: "sin" | "cos" | "tan" | "exp" | "ln" | "sqrt" | "popcount" | "lengthof"
+    cast_operator: declare_type
+    incrementor: "++" | "--
+    pragma: "#pragma {" ... "}"
+    declaration_statement: (quantum_declaration | classical_declaration | constant_declaration) ";"
```

Comment:
Based on the spec, the qubits are global. Should the quantum declaration be added to global statement? This is under "generalities" in https://qiskit.github.io/openqasm/

Reply:
Should there also be a `gate_declaration`?

Reply:
This is already covered in the `gate_definition` in top-level.

Reply:
> What about a statement like: g qr0[0], qr1, qr2[0], qr3; // ok. This would be covered if we make the body optional.

That's a gate call which should be allowed?